rfcs/2020-11-02-remap-infallible-functions.md

```diff
@@ -0,0 +1,371 @@
+not magically make any input valid JSON.
+
+Given the original example, any badly formatted input values are still
+considered invalid JSON and the parser is thus unable to return a representative
+JSON value.
+
+So what should the function return instead?
+
+Infallible functions are allowed to return `Option<Value>`. In other words, they
+can return "nothing", if there isn't anything useful to return.
+
+### How To Handle "Nothing"
+
+Let's take a look at the original example again:
+
+```javascript
+.message = parse_json(.message)
+.timestamp = now()
+```
+
+If `message` contains any data that `parse_json` can't handle, it will return
+`None`.
+
+How that `None` gets processed is up to the expression that processes the value.
+
+For example, in the case of the `Assignment` expression (which is what's called
+when doing `.message = …`), `None` values skip assigning the value to the target
+path, returning `None` itself.
+
+Given this, the outcome of the above example would change from "exception,
+original event untouched", to "JSON parsing failed, `message` field untouched,
+`timestamp` field set to current timestamp".
+
+### Explicitly Handle Error Case
+
+While the above provides more flexibility to the end-user on when to terminate
+the program and when not to, it still does not allow the user to detect if JSON
+parsing failed.
+
+To do this, we modify the `IfStatement` and `Logical` expressions to support
+falling though to the "else" expression if the conditional expression returns
+`None`.
+
+For example:
+
+```javascript
+if ($message = parse_json(.message)) {
+    // parsing succeeded
+} else {
+    // parsing failed
+}
+```
+
+### False vs. "Nothing"
+
+One issue with this proposal is that `false` is a valid JSON value, and so the
+above check does not distinguish between false values or failed JSON parsing.
+
+One solution to this is to _not_ update `IfStatement`, but instead introduce a
+new `is_undefined` function, which returns `true` for `None` values, or else
+returns `false` (or inversely, we add a `is_defined` function):
+
+```javascript
+if ($message = parse_json(.message); is_undefined($message)) {
+    // parsing failed
+} else {
+    // parsing succeeded
+}
+```
+
+It should be noted that this example is flawed, because it expects `$message` to
+be undefined if the parsing failed. Earlier in the examples we documented that
+the assignment expression would not assign to the target if the resolved value
+is `None`. Meaning, if `$message` was already defined earlier in the program, it
+would remain defined to that original value, and the `is_undefined` check would
+return `false`, unexpectedly.
+
+One could avoid this by doing parsing twice:
+
+```javascript
+if parse_json(.message) {
+    // parsing succeeded, do it again!
+    .message = parse_json(.message)
+}
+```
+
+This comes at a performance cost, as the message has to be parsed twice.
+
+An alternative is to consider assigning "nothing" to a target as erasing that
```

Comment:
I think this is preferable as the first example is a pretty big gotcha.

Rather than erased, I would set `.foo` to `null` though. This still has the issue you mention about differentiating a failed parse from parsing the JSON string `null`, but I'll leave an extended comment with a suggested alteration to this proposal that would allow handling of it.